src-d/code-annotation
# DB Migrations

depends on ~~https://github.com/src-d/code-annotation/pull/206~~ and https://github.com/src-d/code-annotation/pull/232

required by https://github.com/src-d/code-annotation/issues/197 and https://github.com/src-d/code-annotation/pull/233

This PR defines the migration from our current Staging DB, with no `uast` cols, to a new internal DB having `uast` nullable cols in the `file_pairs` table as required by https://github.com/src-d/code-annotation/pull/206
This PR also removes the `diff` column from the internal DB as required by https://github.com/src-d/code-annotation/pull/232
This PR also removes the `features` table from the internal DB as required by https://github.com/src-d/code-annotation/pull/233

### How to use it:
1. use an old internal database (`internal.db`) &rarr; can be downloaded from the [export tool](https://code-annotation-staging.srcd.run/export)
1. use a source DB containing uasts (`source.db`) &rarr; [from ML issue](https://github.com/src-d/backlog/issues/1242#issuecomment-372313046), stored [in gdrive](https://drive.google.com/file/d/1J8b9yGOhldMUOmmn6xubPcYMwoQbdTnu/view?usp=sharing)
1. ensure that source DB has the UASTs to import inside a `files` table
1. run the migrations
```shell
# prepares the current "internal.db". Adds "uast_a" and "uast_b" nulable cols
go run cli/migrations/*.go uast-add-cols internal.db
# import UASTs into "internal.db" reading from "source.db"
go run cli/migrations/*.go uast-import internal.db source.db
# prepares the current "internal.db". Remove "diff" col
go run cli/migrations/*.go diff-rm-col internal.db
# drops the 'features' table from the current "internal.db"
go run cli/migrations/*.go features-drop-table internal.db
```

### How to validate the result
with this command/query you can fetch those FilePairs whose UASTs could not be imported
```shell
sqlite3 internal.db 'SELECT blob_id_a, uast_a, blob_id_b, uast_b
 FROM file_pairs
 WHERE uast_a IS NULL or uast_b IS NULL;'
```
if you made a backup of the file_pairs table, you can ensure that primary keys were kept running:
```shell
sqlite3 internal.db 'SELECT count(*)
  FROM file_pairs_backup fb, file_pairs f
  WHERE
    fb.id=f.id and
    fb.blob_id_a=f.blob_id_a and
    fb.blob_id_b=f.blob_id_b and
    fb.uast_a=f.uast_a and
    fb.uast_b=f.uast_b;'
```
The returned value will be the amount of PK that were messed (it should be `0`)

## Disclaimer

**These are not real migrations with up/down feature**

You can read more details about these scripts in the `migrations/README.md`
